### PR TITLE
[ISSUE #3484] fix springcloud client lossless registration

### LINKED ISSUE #3484: [Task] client lossless registration
### Description

At present, almost all clients(exclude `dubbo`) have a problem: when the spring container has not been started, it reports the registration information to `shenyu-admin`. If there is traffic coming, the downstream system call will fail directly (for RPC clients, it will report an error such as `no provider`; for HTTP clients, it will fail directly, or the call times out), which also leads to `lossy publishing`.

The root cause of this problem is that the registration of the client depends on the `BeanPostProcessor`, so we need to change it to a form based on `ContextRefreshedEvent`(you can see it in `org.apache.shenyu.client.apache.dubbo.ApacheDubboServiceBeanListener`).

Check https://github.com/apache/incubator-shenyu/issues/415 for more details about `How dubbo client solve this`.

### Task List

1. - [x] grpc client lossless registration; @renzhuyan 
2. - [x] springmvc client lossless registration; @gaochuang98 
3. - [x] springcloud client lossless registration; @renzhuyan 
4. - [x] motan client lossless registration; @ShawnSiao 
5. - [x] sofa client lossless registration; @lahmXu 
6. - [x] tars client lossless registration; @gaochuang98 
7. - [x] websocket client lossless registration; @lahmXu

### PR DESCRIPTION
**Pull request by renzhuyan**
Make sure that:

- [ ] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [ ] You submit test cases (unit or integration tests) that back your changes.
- [ ] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java [hunk 1]
```diff
@@ -46,21 +46,22 @@
...
 import java.util.Objects;
 import java.util.Optional;
 import java.util.Properties;
 
 /**
- * The type Shenyu client bean post processor.
+ * The type Shenyu client bean listener.
  */
-public class SpringCloudClientBeanPostProcessor implements BeanPostProcessor {
+public class SpringCloudClientBeanListener implements ApplicationListener<ContextRefreshedEvent> {
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java:57]: please rename to `SpringCloudClientEventListener`.

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java [hunk 2]
```diff
@@ -101,24 +102,34 @@ public SpringCloudClientBeanPostProcessor(final PropertiesConfig clientConfig,
...
         publisher.start(shenyuClientRegisterRepository);
     }
-    
+
     @Override
-    public Object postProcessAfterInitialization(@NonNull final Object bean, @NonNull final String beanName) throws BeansException {
+    public void onApplicationEvent(final ContextRefreshedEvent contextRefreshedEvent) {
+        if (Objects.nonNull(contextRefreshedEvent.getApplicationContext().getParent())) {
+            return;
+        }
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java:110]: What is the point of doing this?
```java
if (Objects.nonNull(contextRefreshedEvent.getApplicationContext().getParent())) {
    return;
}
```

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java [hunk 3]
```diff
@@ -89,12 +91,12 @@ public void testWithShenyuClientAnnotation() {
         registerUtilsMockedStatic.when(() -> RegisterUtils.doLogin(any(), any(), any())).thenReturn(Optional.of("token"));
         registerUtilsMockedStatic.when(() -> RegisterUtils.doRegister(any(), any(), any()))
                 .thenAnswer((Answer<Void>) invocation -> null);
-        SpringCloudClientBeanPostProcessor springCloudClientBeanPostProcessor = buildSpringCloudClientBeanPostProcessor(false);
-        assertThat(springCloudClientTestBean, equalTo(springCloudClientBeanPostProcessor.postProcessAfterInitialization(springCloudClientTestBean, "normalBean")));
+        SpringCloudClientBeanListener springCloudClientBeanListener = buildSpringCloudClientBeanPostProcessor(false);
+        springCloudClientBeanListener.onApplicationEvent(contextRefreshedEvent);
         registerUtilsMockedStatic.close();
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java:96]: missed `assert` judge.

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java [hunk 4]
```diff
@@ -58,29 +58,31 @@ public final class SpringCloudClientBeanPostProcessorTest {
...
 
     @Test
     public void testShenyuBeanProcess() {
         registerUtilsMockedStatic.when(() -> RegisterUtils.doLogin(any(), any(), any())).thenReturn(Optional.of("token"));
         // config with full
-        SpringCloudClientBeanPostProcessor springCloudClientBeanPostProcessor = buildSpringCloudClientBeanPostProcessor(true);
-        assertThat(springCloudClientTestBean, equalTo(springCloudClientBeanPostProcessor.postProcessAfterInitialization(springCloudClientTestBean, "springCloudClientTestBean")));
+        SpringCloudClientBeanListener springCloudClientBeanListener = buildSpringCloudClientBeanPostProcessor(true);
+        springCloudClientBeanListener.onApplicationEvent(contextRefreshedEvent);
         registerUtilsMockedStatic.close();
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java:78]: missed assert judge.

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java (hunk 1 above)
Comment (renzhuyan) [reply shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java]: ok, i renamed.

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java (hunk 2 above)
Comment (renzhuyan) [reply shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/main/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanListener.java]: I will remove

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java (hunk 3 above)
Comment (renzhuyan) [reply shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java:108]: Method has no return value. I added the call count vetify.

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java (hunk 4 above)
Comment (renzhuyan) [reply shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java:88]: Method has no return value. I added the call count vetify.

File: shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java [hunk 5]
```diff
@@ -89,12 +102,13 @@ public void testWithShenyuClientAnnotation() {
...
-        SpringCloudClientBeanPostProcessor springCloudClientBeanPostProcessor = buildSpringCloudClientBeanPostProcessor(false);
-        assertThat(springCloudClientTestBean, equalTo(springCloudClientBeanPostProcessor.postProcessAfterInitialization(springCloudClientTestBean, "normalBean")));
+        SpringCloudClientEventListener springCloudClientEventListener = buildSpringCloudClientBeanPostProcessor(false);
+        springCloudClientEventListener.onApplicationEvent(contextRefreshedEvent);
+        verify(applicationContext, times(1)).getBeansWithAnnotation(any());
         registerUtilsMockedStatic.close();
     }
 
-    private SpringCloudClientBeanPostProcessor buildSpringCloudClientBeanPostProcessor(final boolean full) {
+    private SpringCloudClientEventListener buildSpringCloudClientBeanPostProcessor(final boolean full) {
```
Comment (dragon-zhang) [inline shenyu-client/shenyu-client-http/shenyu-client-springcloud/src/test/java/org/apache/shenyu/client/springcloud/init/SpringCloudClientBeanPostProcessorTest.java:111]: rename to `buildSpringCloudClienttEventListener`.

**Review by dragon-zhang — APPROVED** [review]

**Review by dragon-zhang — APPROVED** [review]

**Pull request merged**